packages/backend/server/src/plugins/copilot/providers/perplexity.ts

```diff
@@ -0,0 +1,209 @@
+import {
+  CopilotCapability,
+  CopilotChatOptions,
+  CopilotProviderType,
+  CopilotTextToTextProvider,
+  PromptMessage,
+} from '../types';
+
+export type PerplexityConfig = {
+  apiKey: string;
+};
+
+interface Message {
+  role: 'assistant';
+  content: string;
+}
+
+interface Choice {
+  message: Message;
+  delta: Message;
+  finish_reason: 'stop' | null;
+}
+
+interface PerplexityData {
+  citations: string[];
+  choices: Choice[];
+}
+
+export function injectCitations(content: string, citations: string[]) {
+  // Match [[n]] and [n] patterns
+  // Not match if they're already part of a formatted citation
+  const regex = /(?:\[\[(\d+)\]\]|\[(\d+)\])(?!\]?\([^)]*\))/g;
+  return content.replace(regex, (_, g1, g2) => {
+    const index = parseInt(g1 || g2) - 1;
+    if (index >= 0 && index < citations.length) {
+      return `[[${g1 || g2}](${citations[index]})]`;
+    }
+    return _;
+  });
+}
+
+export class PerplexityProvider implements CopilotTextToTextProvider {
+  static readonly type = CopilotProviderType.Perplexity;
+
+  static readonly capabilities = [CopilotCapability.TextToText];
+
+  static assetsConfig(config: PerplexityConfig) {
+    return !!config.apiKey;
+  }
+
+  constructor(private readonly config: PerplexityConfig) {
+    assert(PerplexityProvider.assetsConfig(config));
+  }
+
+  readonly availableModels = [
+    'llama-3.1-sonar-small-128k-online',
+    'llama-3.1-sonar-large-128k-online',
+    'llama-3.1-sonar-huge-128k-online',
+  ];
+
+  get type(): CopilotProviderType {
+    return PerplexityProvider.type;
+  }
+
+  getCapabilities(): CopilotCapability[] {
+    return PerplexityProvider.capabilities;
+  }
+
+  async isModelAvailable(model: string): Promise<boolean> {
+    return this.availableModels.includes(model);
+  }
+
+  async generateText(
+    messages: PromptMessage[],
+    model: string = 'llama-3.1-sonar-small-128k-online',
+    options: CopilotChatOptions = {}
+  ): Promise<string> {
+    await this.checkParams({ messages, model, options });
+    try {
+      metrics.ai.counter('chat_text_calls').add(1, { model });
+      const sMessages = messages
+        .map(({ content, role }) => ({ content, role }))
+        .filter(({ content }) => typeof content === 'string');
+
+      const params = {
+        method: 'POST',
+        headers: {
+          Authorization: `Bearer ${this.config.apiKey}`,
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify({
+          model,
+          messages: sMessages,
+          max_tokens: options.maxTokens || 4096,
+        }),
+      };
+      const response = await fetch(
+        'https://api.perplexity.ai/chat/completions',
+        params
+      );
+      const json: PerplexityData = await response.json();
+      return injectCitations(json.choices[0].message.content, json.citations);
+    } catch (e: any) {
+      metrics.ai.counter('chat_text_errors').add(1, { model });
+      throw this.handleError(e);
```

Comment:
better to add `params` into the error or add it to log so we can debug the `unexpected_response` situation